[SPARK-12941][SQL][BRANCH-1.4] Spark-SQL JDBC Oracle dialect fails to map string datatypes to Oracle VARCHAR datatype

thomastechs opened the pull request:
Adding a getJDBCType() method to the JdbcDialects.scala which would create a VARCHAR type for Oracle create table operations.Currently the type TEXT is throwing exceptions which is incompatible to Oracle

thomastechs committed:
Update for [SPARK-12941][SQL]

Adding a getJDBCType() method to the JdbcDialects.scala which would create a VARCHAR type for Oracle create table operations.Currently the type TEXT is throwing exceptions which is incompatible to Oracle

thomastechs commented:
@yhuai     fyr

AmplabJenkins commented:
Can one of the admins verify this patch?

yhuai commented:
Can you also create a pr for master? 

btw, it is possible to add a docker integration test (like MySQLIntegrationSuite)?

yhuai commented:
test this please

SparkQA commented:
[Test build #50107 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50107/consoleFull) for   PR 10912 at commit [`a479fce`](https://github.com/apache/spark/commit/a479fcee25c12b8f4ff32ee732ad7cdc3a73398c).

SparkQA commented:
[Test build #50107 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50107/consoleFull) for   PR 10912 at commit [`a479fce`](https://github.com/apache/spark/commit/a479fcee25c12b8f4ff32ee732ad7cdc3a73398c).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/50107/
Test FAILed.

yhuai commented:
please fix the style error

thomastechs committed:
Update JdbcDialects.scala

Updated the style error-remove whitespace

thomastechs commented:
@yhuai  The style check error has been fixed and committed.

yhuai commented:
Can you add a test? Also, can you create a pr for master?

thomastechs committed:
Update for [SPARK-12941][SQL] JDBCSuite.scala

Adding test case for the getJDBCType() method for the datatype-StringType as a part of OracleDialect

thomastechs commented:
@yhuai  Added the testcase for 1.4.     For the master PR, would it be good to raise another bug in reference to this and upload the fix as a new PR for that bug.? In that case, we could close this quickly (once tests passed) , and release this fix for Jose who raised this bug, for his purpose.

yhuai commented on sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala:
```diff
@@ -445,4 +445,9 @@ class JDBCSuite extends SparkFunSuite with BeforeAndAfter {
+  test("OracleDialect type mapping") {
+    val oracleDialect = JdbcDialects.get("jdbc:oracle://127.0.0.1/db")
+    assert(oracleDialect.getJDBCType(StringType).map(_.databaseTypeDefinition).get == "VARCHAR2(255)")
+  }
```
I am not sure this test works. Let me be clear. We need to add a docker integration test like `MySQLIntegrationSuite`. Since spark 1.4 branch does not have the testing infrastructure. I think the the right way to do it is for your new PR (using the same JIRA id in the title) created for master, you add the test. Then, for this PR, we just keep the fix.

yhuai commented:
Please create a PR against master. You do not need to create a new jira. In the title of the new PR, you just use the same JIRA number. For this PR, it is better to add `[BRANCH-1.4]` in the title.

yhuai commented:
btw, just want to double check. Have you tested your fix with Oracle?

yhuai commented:
Also, I'd like to get the change in master first. Thanks.

thomastechs commented:
@yhuai  Yes;tested.  So, as you mentioned, will create a new PR against master with the OracleIntegrationSuite.

yhuai commented:
test this please

SparkQA commented:
[Test build #50199 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50199/consoleFull) for   PR 10912 at commit [`06bb080`](https://github.com/apache/spark/commit/06bb0805e221ccc66aa29805e9fac139f6343efc).

SparkQA commented:
[Test build #50199 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50199/consoleFull) for   PR 10912 at commit [`06bb080`](https://github.com/apache/spark/commit/06bb0805e221ccc66aa29805e9fac139f6343efc).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/50199/
Test FAILed.

thomastechs commented:
@yhuai   Can the below line wrap pass the style check?
    assert(oracleDialect.getJDBCType(StringType)
      .map(_.databaseTypeDefinition).get == "VARCHAR2(255)")

thomastechs commented:
Read it as 2 spaces before .map

yhuai commented:
If you look at the link, you will see (https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50199/consoleFull)
```
Scalastyle checks failed at following occurrences:
[error] /home/jenkins/workspace/SparkPullRequestBuilder/sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala:450: File line length exceeds 100 characters
[error] (sql/test:scalastyle) errors exist
[error] Total time: 7 s, completed Jan 27, 2016 10:12:08 AM
[error] /home/jenkins/workspace/SparkPullRequestBuilder/sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala:450: File line length exceeds 100 characters
[error] (sql/test:scalastyle) errors exist
[error] Total time: 6 s, completed Jan 27, 2016 10:12:47 AM
[error] Got a return code of 1 on line 133 of the run-tests script.
```

Also, I do not think the test in this PR will work. Please put that in the OracleIntegrationSuite in the PR for master.

thomastechs commented:
@yhuai  working on the OracleIntgerationSuite  (In the process of docker build using Oracle linux)

thomastechs commented:
@yhuai  and @JoshRosen 
Sorry for the delay. I was working with @jayadevanmurali to bring up the docker based test suite for master branch.
We are able to create the test case with OracleIntegrationSuite and successfully tested this fix in our local environment.The version we used is Oracle 11g:, and used the docker image from wnameless/oracle-xe-11g.
We tried to build using maven repository for the ojdbc jar required for this fix.Added the dependency like below in the pom file (docker integration project).
   <dependency>
            <groupId>com.oracle</groupId>
            <artifactId>ojdbc6</artifactId>
            <version>11.2.0.2.0</version>
           <scope>test</scope>
    </dependency>
But there was no corresponding ojdbc jar in the maven repository. So we had to manually download it to the local and build it.
So, if we create a pull request, how these jar to be uploaded?

thomastechs commented:
@yhuai 
Created the PR https://github.com/apache/spark/pull/11306/files for the master branch

thomastechs committed:
Update JDBCSuite.scala [SPARK-12941][Branch-1.4]

Updated for the style fix.

thomastechs commented:
@yhuai    The style check error for this PR has been fixed. This is for branch 1.4

yhuai commented on sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala:
```diff
@@ -445,4 +445,10 @@ class JDBCSuite extends SparkFunSuite with BeforeAndAfter {
+    assert(oracleDialect.getJDBCType(StringType).
+      map(_.databaseTypeDefinition).get == "VARCHAR2(255)")
+  }
+
```
This test does not work, right? If so, can we remove it?

thomastechs replied:
This is similar to testcase written for MySQL,for branch 1.4; which is not actually spinning up any MySQL or Oracle instance; but for the test suite purpose, as I believe.  So far, builds for this PR for branch 1.4, failed due to style checks.

thomastechs committed:
Update JDBCSuite.scala

yhuai commented:
test this please

SparkQA commented:
[Test build #52189 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52189/consoleFull) for   PR 10912 at commit [`b46f1b3`](https://github.com/apache/spark/commit/b46f1b3c2c6ca43077ad77bd90d05ae5c1d9923e).

yhuai commented on sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala:
```diff
@@ -445,4 +445,9 @@ class JDBCSuite extends SparkFunSuite with BeforeAndAfter {
+    val oracleDialect = JdbcDialects.get("jdbc:oracle://127.0.0.1/db")
+    assert(oracleDialect.getJDBCType(StringType).
+      map(_.databaseTypeDefinition).get == "VARCHAR2(255)")
+  }
```
OK I see. Originally, I thought it requires a oracle db. Thank you for the explanation.

Then, we should also add this test in the master. Can you submit a PR? We can reuse the jira number.

thomastechs replied:
Okei @yhuai    So I shall submit another PR, for the same JIRA with the updates in the JDBCSuite.scala in the master branch. I shall also close this PR

SparkQA commented:
[Test build #52189 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52189/consoleFull) for   PR 10912 at commit [`b46f1b3`](https://github.com/apache/spark/commit/b46f1b3c2c6ca43077ad77bd90d05ae5c1d9923e).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/52189/
Test PASSed.

yhuai commented on sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala:
```diff
@@ -445,4 +445,9 @@ class JDBCSuite extends SparkFunSuite with BeforeAndAfter {
     assert(agg.getCatalystType(1, "", 1, null) == Some(StringType))
   }
 
+  test("OracleDialect type mapping") {
```
Also, it will be good to make the test name more specific on what it will test. I am going to change it while merging.

yhuai commented:
Merging to 1.4 and 1.5

yhuai commented:
ok I have merged it into branch 1.4 and branch 1.5. Can you close this PR? Thanks!
